File: src/hardware/gpio-interface/GPIOInterface.cpp
#include "include/rocko_env/GPIOInterface.hpp"


namespace rocko_env
{
    bool GPIOInterface::setupPin(string pinName, bool isOut) {
        // Call the function
        PyObject *result = PyObject_CallFunction(setupPinFunc, "sb", pinName.c_str(), isOut);
        if (!PyLong_Check(result)) {
            PyErr_Print();
            return false;
        }
        return PyLong_AsLong(result) == 0;
    }

    bool GPIOInterface::startPWM(string pinName, int dutyCycle, int freq, bool isFallingEdge) {
        PyObject *result = PyObject_CallFunction(startPWMFunc, "siib", pinName.c_str(), dutyCycle, freq, isFallingEdge);
        if (!PyLong_Check(result)) {
            PyErr_Print();
            return false;
        }
        return PyLong_AsLong(result) == 0;
    }

    bool GPIOInterface::stopPWM(string pinName) {
        PyObject *result = PyObject_CallFunction(stopPWMFunc, "s", pinName.c_str());
        if (!PyLong_Check(result)) {
            PyErr_Print();
            return false;
        }
        return PyLong_AsLong(result) == 0;
    }
        
    bool GPIOInterface::setDutyCycle(string pinName, int dutyCycle) {
        PyObject *result = PyObject_CallFunction(setDutyCycleFunc, "si", pinName.c_str(), dutyCycle);

        if (!PyLong_Check(result)) {
            PyErr_Print();
            return false;
        }
        return PyLong_AsLong(result) == 0;
    }
}

File: src/hardware/gpio-interface/gpio_interface.py
# import Adafruit_BBIO.GPIO as GPIO
# import Adafruit_BBIO.PWM as PWM

def setup_pin(pin_name:str, is_out:bool):
    # dir = GPIO.OUT if is_out else GPIO.IN
    # GPIO.setup(str, dir)
    return 0

def start_pwm(pin_name:str, duty_cycle:int, freq:int, is_falling_edge:bool):
    # PWM.start(pin_name, duty_cycle, freq, is_falling_edge)
    return 0

def stop_pwm(pin_name:str):
    # PWM.stop(pin_name)
    return 0

def set_duty_cycle(pin_name:str, duty_cycle:int):

    # PWM.set_duty_cycle(pin_name, duty_cycle)
    return 0

